examples/hf_serialization.py

```diff
@@ -226,71 +303,86 @@ def df_main(args: argparse.Namespace) -> None:
+    for prefix, component in (
+        ("encoder", pipeline.text_encoder),
+        ("vae", pipeline.vae),
+        ("unet", pipeline.unet),
+    ):
+        component.eval()
+        config = component.config
+        if isinstance(config, dict):
+            config = {k: v for k, v in config.items() if k != "_name_or_path"}
+        serialize_model(
+            component, config, output_prefix, prefix, force=args.force
+        )
 
-    pipeline.tokenizer.save_pretrained(output_prefix)
-    pipeline.scheduler.save_pretrained(output_prefix)
+    serialize_pretrained(
+        pipeline.tokenizer, output_prefix, "tokenizer", force=args.force
+    )
+    serialize_pretrained(
+        pipeline.scheduler, output_prefix, "scheduler", force=args.force
+    )
+    tokenizer_type = type(pipeline.tokenizer)
+    scheduler_type = type(pipeline.scheduler)
 
     if args.validate:
-        del pipeline
-        gc.collect()
         device = utils.get_device()
+        pipeline = pipeline.to(device)
+        gc.collect()
 
         logger.info("Validating serialization")
         vae = load_model(
-            output_prefix, AutoencoderKL, None, "vae", device, hf_id=hf_id
+            output_prefix,
+            AutoencoderKL,
+            None,
+            "vae",
+            device,
         )
+        assert_module_equal(pipeline.vae, vae, "vae")
+
         unet = load_model(
             output_prefix,
             UNet2DConditionModel,
             None,
             "unet",
             device,
-            hf_id=hf_id,
         )
+        assert_module_equal(pipeline.unet, unet, "unet")
+
         encoder = load_model(
             output_prefix,
             CLIPTextModel,
             CLIPTextConfig,
             "encoder",
             device,
-            hf_id=hf_id,
         )
+        assert_module_equal(pipeline.text_encoder, encoder, "encoder")
+
+        tokenizer = load_pretrained(tokenizer_type, output_prefix, "tokenizer")
+        scheduler = load_pretrained(scheduler_type, output_prefix, "scheduler")
 
-        pipeline = StableDiffusionPipeline(
+        deserialized_pipeline = StableDiffusionPipeline(
             text_encoder=encoder,
             vae=vae,
             unet=unet,
-            tokenizer=AutoTokenizer.from_pretrained(
-                args.input_directory, subfolder="tokenizer"
-            ),
-            scheduler=LMSDiscreteScheduler.from_pretrained(
-                args.input_directory, subfolder="scheduler"
-            ),
+            tokenizer=tokenizer,
+            scheduler=scheduler,
             safety_checker=None,
             feature_extractor=None,
             requires_safety_checker=False,
         ).to(device)
 
         prompt = "a photo of an astronaut riding a horse on mars"
         with torch.autocast(device.type):
-            pipeline(prompt).images[0].save("test.png")
+            deserialized_pipeline(prompt).images[0].save("test.png")
```

Comment:
I would get rid of the test inference code since the previous validation steps should suffice for verifying that the model weights are unchanged.

Reply:
Commented it out in 032a12ff0b3e77ac724b719b38565096e0996580

Reply:
Also added validation for the deserialized tokenizer (and a small amount of validation for the deserialized scheduler) to compensate here, since this was the only part making use of those two components.